llvm/lib/Transforms/Yk/BasicBlockTracer.cpp

```diff
@@ -0,0 +1,72 @@
+#include "llvm/Transforms/Yk/BasicBlockTracer.h"
+#include "llvm/IR/BasicBlock.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instruction.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/LLVMContext.h"
+#include "llvm/IR/Module.h"
+#include "llvm/InitializePasses.h"
+#include "llvm/Pass.h"
+
+#define DEBUG_TYPE "yk-basicblock-tracer-pass"
+
+using namespace llvm;
+
+namespace llvm {
+void initializeYkBasicBlockTracerPass(PassRegistry &);
+} // namespace llvm
+
+namespace {
+struct YkBasicBlockTracer : public ModulePass {
+  static char ID;
+  Function *externalFunc = NULL;
+
+  YkBasicBlockTracer() : ModulePass(ID) {
+    initializeYkBasicBlockTracerPass(*PassRegistry::getPassRegistry());
+  }
+
+  bool doInitialization(Module &M) override {
+    LLVMContext &Context = M.getContext();
+    if (externalFunc == NULL) {
+      // Create externally linked function declaration:
+      //   void yk_trace_basicblock(int functionIndex, int blockIndex)
+      Type *returnType = Type::getVoidTy(Context);
+      Type *functionIndexArgType = Type::getInt32Ty(Context);
+      Type *blockIndexArgType = Type::getInt32Ty(Context);
+
+      FunctionType *FType = FunctionType::get(
+          returnType, {functionIndexArgType, blockIndexArgType}, false);
+      externalFunc = Function::Create(FType, GlobalVariable::ExternalLinkage,
+                                      YK_TRACE_FUNCTION, M);
+      return true;
+    }
+    return false;
+  }
+
+  bool runOnModule(Module &M) override {
+    int functionIndex = 0;
+    IRBuilder<> builder(M.getContext());
+    for (auto& F : M) {
+      int blockIndex = 0;
+      for (auto& BB : F) {
+        builder.SetInsertPoint(&*BB.getFirstInsertionPt());
+        builder.CreateCall(externalFunc, {builder.getInt32(functionIndex),
```

Comment:
Is `i32` sufficient? I think so, because LLVM uses `unsigned` for most indices, which should cover us for the arches we care for.

Maybe an assertion to double check?

Reply:
Should we use the same type as LLVM? `i32` and `unsigned int` aren't always the same thing.

Reply:
I can change BlockIndex and FunctionIndex to be `unsigned`.

In LLVM IRBuilder `getInt32` seems to be just about 32 bits. It looks like it can be both signed and unsigned?

```cpp
  /// Get a constant 32-bit value.
  ConstantInt *getInt32(uint32_t C) {
    return ConstantInt::get(getInt32Ty(), C);
  }
```
> Maybe an assertion to double check?

what kind of assertion?